### ISSUE TITLE
Don't use EscapeUriString for Redirections

### ISSUE THREAD

**Issue by jtkech** [OrchardCMS/OrchardCore#10383]
@sebastienros 

Yes, it will be obsolete but in the meantime maybe we should not use it (see below)

They suggest `Uri.EscapeDataString()` but it escapes more chars, e.g. the reserved `?` of a query string as we use e.g. to maintain the state of the admin menu. Not easy to escape an url that may contain or not different uri components with their own reserved chars as for a queryString, and so on.

So I started to work on it but at some point I did a test, to not have to change the path I opted to use an admin menu item with an additional query param `tété=tété`, the unescaped and escaped queryString of the returnUrl becoming.

    ?admin=-1147159492&tété=tété
    ?admin=1280032454&t%C3%A9t%C3%A9=t%C3%A9t%C3%A9

Then while editing a content item, because the returnUrl itself is passed as a parameter, the whole returnUrl is **url encoded again** as any other parameter would be.

    %3Fadmin%3D1280032454%26t%25C3%25A9t%25C3%25A9%3Dt%25C3%25A9t%25C3%25A9

And when the controller action is hit, the binder url decodes it once and we retrieve what we passed (encoded or not).

    ?admin=-1147159492&t%C3%A9t%C3%A9=t%C3%A9t%C3%A9

Finally if we edit / save multiple times a content item and return back multiple times to the admin list, if we don't use `Uri.EscapeUriString()` the above queryString remains the same which is okay. But if we use `Uri.EscapeUriString()` the queryString part of the returnUrl is encoded once too often each time, resulting in the following.

    ?admin=-1147159492&t%C3%A9t%C3%A9=t%C3%A9t%C3%A9
    ?admin=-1147159492&t%25C3%25A9t%25C3%25A9=t%25C3%25A9t%25C3%25A9
    ?admin=-1147159492&t%2525C3%2525A9t%2525C3%2525A9=t%2525C3%2525A9t%2525C3%2525A9
    // and so on

So I think that in the controller we always retrieve the original returnUrl, maybe we should not use `Uri.EscapeUriString()` for a redirection, and if it doesn't work it would mean that this is the original returnUrl that is not well escaped (at least one of its uri component)

So maybe we should revert #10368.

**Comment by sebastienros** [comment]
In general I don't see how there could be an issue if everything that needs to be encoded is encoded. If there is an issue in the process you describe, then some assumptions are wrong and need to be fixed, meaning some decoding/encoding are missing or extraneous.

Is the case you describe here the only one you have found with a problem?

**Comment by sebastienros** [comment]
`EscapeDataString` escapes too much :/

> I opted to use an admin menu item with an additional query param

What is a repro of the issue with `EscapeUriString`? I don't understand this one.

**Comment by sebastienros** [comment]
Currently testing this method instead, it correctly handles relative and absolute urls, and only escaped what needs to be:

```
"/login/returnUrl=%20té%C3%A9" => /login/returnUrl=%20t%C3%A9%C3%A9
"https://foo/login/returnUrl=%20té%C3%A9" => https://foo/login/returnUrl=%20t%C3%A9%C3%A9
```

```
static string EscapeLocationHeader(string path)
{
    var uri = new Uri(path, UriKind.RelativeOrAbsolute);
    return uri.GetComponents(UriComponents.SerializationInfoString, UriFormat.UriEscaped);
}
```

**Comment by jtkech** [comment]
@sebastienros 

So yes, yesterday to go quick I used the admin menu to create an admin menu link with an additional param, this to confirm that the binded returnUrl is already escaped, and that `EscapeUriString()` does extra encoding.

    /Admin/Contents/ContentItems/?tété=tété

Today, just tried to modify the coded route of the content items admin list with extra `éé`.

    /Admin/Contents/ContentItemséé/{contentTypeId?}

Then if you edit and save a content item, the controller see the following escaped returnUrl

    /Admin/Contents/ContentItems%C3%A9%C3%A9?admin=475673731

But then `EscapeUriString()` does an extra escaping and it just fails when redirecting to the admin list.

    /Admin/Contents/ContentItems%25C3%25A9%25C3%25A9?admin=-137263071

Yes we could try to find a way to correctly escape each uri component (each one may have e.g. their own reserved chars), and **only if not already escaped**. E.g. there are some helpers that escape a query string param but only if not already done. E.g. `EscapeUriString()` calls an internal helper with a `checkExistingEscaped` param that would allow this but that is internally set to false.

But ATM I think that it is better to do nothing here and then just ensure that the passed returnUrl is always well escaped, by our code e.g. by using the url helper or just because the browser does it for us (hmm, here not sure when this is a script that manages the queryString through an ajax call). So if there are some use case where it doesn't work, maybe some extraneous encoding of the returnUrl before passing it as a query string parameter, I will try the use case of @urbanit.

**Comment by urbanit** [comment]
So, redirect url should be properly escaped. 
Q: is there any safe way to identify this case?
If there is, we should verify it before redirection and escape only if needed.
If there is not a safe way, it's a l-l situation... Either no escape while needed (responsibility of admin to pass escaped urls) or escape already escaped URLs...

**Comment by jtkech** [comment]
Maybe #10392 will fix it by escaping but only if needed

But I will try to repro your use case to check that there is still not an extra encoding (not url related)

**Comment by jtkech** [comment]
Okay I start to understand

Cool to be able to edit from the front end ;)

So I tried it by using an article with `/abouté` autoroute path

The liquid generates an href with the following query string

    <a href="/.../Edit?returnUrl=/about%C3%A9" class="btn btn-primary btn-sm">Edit</a>

Okay the `returnUrl` is well escaped as an url returned by the `display_url` filter

But to pass it as a parameter of the queryString of the main href, we need to url encoded it again as for any other parameters we would want to pass.

Based on my tests would need to be something like the following where the `%` itself is escaped to `%25`.

    <a href="/.../Edit?returnUrl=/about%25C3%25A9" class="btn btn-primary btn-sm">Edit</a>


Otherwise when the binder unescapes it, in the controller we retrieve `/abouté` which is not a well escaped url.

We would need an `url_encode` liquid filter to encode query string parameters.

**Comment by jtkech** [comment]
When we use form / input helpers it works as the returnUrl is encoded again to be passed as a bindable parameter, e.g.

        @Html.Hidden("returnUrl", Context.Request.Query["returnUrl"])

OKAY, in liquid the following works


    {% assign allowEdit = User | has_permission:"EditContent", Model.ContentItem %}
    {% if allowEdit %}

    <form action="/Admin/Contents/ContentItems/{{ Model.ContentItem.ContentItemId }}/Edit" method="get">
        <button type="submit" class="btn btn-primary btn-sm">Edit</button>
        <input id="returnUrl" name="returnUrl" type="hidden" value="{{ Model.ContentItem | display_url }}">
    </form>

    {% endif %}

**Comment by jtkech** [comment]
closed with #10392
